specification/dartLangSpec.tex

```diff
@@ -11829,7 +11829,40 @@ \subsection{Function Expressions}
 
 \rationale{%
 This definition guarantees that for any type $T$,
-\code{$T <:$ FutureOr<$\flatten{T}$>}.%
+\code{$T <:$ FutureOr<$\flatten{T}$>}. The proof is by induction on `T`:
+
+\begin{itemize}
+\item If $T$ is \code{$X$\,\&\,$S$} then
+  \begin{itemize}
+  \item if $S$ derives a future type $U$,
+    then \code{$T <: S$} and \code{$S <: U$}, so \code{$T <: U$}.
+    By the induction hypothesis, \code{$U <:$ FutureOr<$\flatten{U}$>}.
+    Since \code{$\flatten{T} = \flatten{U}$} in this case, it follows that
+    \code{$U <:$ FutureOr<$\flatten{T}$>}, and so
+    \code{$T <:$ FutureOr<$\flatten{T}$>}.
+  \item otherwise, \code{$T <: X$}.
+    By the induction hypothesis, \code{$X <:$ FutureOr<$\flatten{X}$>}.
+    Since \code{$\flatten{T} = \flatten{X}$} in this case, it follows that
+    \code{$U <:$ FutureOr<$\flatten{T}$>}, and so
+    \code{$T <:$ FutureOr<$\flatten{T}$>}.
+  \end{itemize}
+
+\item If $T$ derives a future type \code{Future<$S$>}
+  or \code{FutureOr<$S$>}, then, since \code{Future<$S$> $<:$ FutureOr<$S$>},
+  it follows that \code{$T <:$ FutureOr<$S$>}. Since \code{$\flatten{T} = S$}
+  in this case, it follows that \code{$T <:$ FutureOr<$\flatten{T}$>}.
+
+\item If $T$ derives a future type \code{Future<$S$>?} or
+  \code{FutureOr<$S$>?}, then, since \code{Future<$S$>? $<:$ FutureOr<$S$>?},
+  it follows that \code{$T <:$ FutureOr<$S$>?}, and therefore
+  \code{$T <:$ FutureOr<$S$?>}. Since \code{$\flatten{T} = S$?} in this case,
```

Comment:
... because `FutureOr<T>? <: FutureOr<T?>` for any type `T`, which can be shown using the union type subtype rules and from `Future<T> <: Future<T?>` by covariance.

Reply:
Good point, this was a reasoning gap in my proof. Updated!